### ISSUE TITLE
Single-file: Facilitate signing on mac

### ISSUE THREAD

**Issue by swaroop-sridhar** [dotnet/runtime#3671]
The single-file bundler adds additional contents (the apps's dependencies) at the end of the host-binary. This violates the strict consistency checks performed by codesign tool in MAC. 

This work item track the effort to make the MAC bundle amenable to signing.

#### Original issue

https://github.com/dotnet/core/issues/2811 

#### Repro stesps 

1. Create a console "Hello world" executable on macOS. 
2. Generate a single file executable followed: https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0#single-file-executables
3. run codesign with this exe. 

You will get error message: main executable failed strict validation

If you add --no-strict option in codesign, you will get more detailed error messages:
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/codesign_allocate: the __LINKEDIT segment does not cover the end of the file (can't be processed) in: /Volumes/Data/temp/HelloDotNet/bin/Release/netcoreapp3.0/osx.10.12-x64/publish/MacOS/HelloDotNet
/Volumes/Data/temp/HelloDotNet/bin/Release/netcoreapp3.0/osx.10.12-x64/publish/MacOS/HelloDotNet: the codesign_allocate helper tool cannot be found or used
```

**Comment by richardlalancette** [comment]
ok, this is going to block distribution of the app. Not cool.

**Comment by richardlalancetteyoui** [comment]
What is my alternative to start sharing my dotnet 3.1 application with my clients?

**Comment by swaroop-sridhar** [comment]
@richardlalancette  You can distribute your app without publishing it as a single-file. The single-file-ness of apps is less important on MAC, sine MAC treats apps as folders.

**Comment by richardlalancetteyoui** [comment]
Oh that's a good point!

**Comment by ProbablyFaiz** [comment]
Any timeline on this? Seems like a pretty urgent problem that proper distribution on one platform is completely blocked.

**Comment by morphx666** [comment]
> 
> 
> The fix is somewhat fragile in the long run, because it relies on the current behavior and checks implemented in `codesign` tool. It may break if the tool introduces more checks (ex: check that all bytes in the string table are referenced).
> 
> The reliable way to generate an image is to use the native tool chain, but this involves substantial support in and beyond the .net core SDK.

But isn't .NET Core supposed to be cross-platform? Isn't that one of its major selling points?
How are we supposed to compile for macOS if we can't sign the binaries?

**Comment by morphx666** [comment]
I can confirm that, whenever selecting the option to generate a single-file , the generated binary cannot be signed.

Is this going to be fixed?
What's the point of having a cross-platform framework if the generated binaries cannot be signed for distribution?

**Comment by agocke** [comment]
Apple has a variety of restrictions on exactly what needs to happen for signing. Those restrictions change over time and we try to address them for various deployment models. Single-file embedding is a new deployment model that will need new work to deal with additional signing restrictions. This probably won't make .NET 5, but there are a variety of other deployment models that are signable, including the existing self-contained deployment, that will continue to work for .NET 5.

Is there a particular issue that's blocking you from using self-contained deployment on macOS right now?

**Comment by ProbablyFaiz** [comment]
Single file is much smaller to distribute, which is very helpful for desktop applications running .NET Core (which I understand is not a priority, but it would be nice).

**Comment by agocke** [comment]
If the main concern is size, you can enable trimming even without enabling single file. There are some single-file specific size improvements, but AFAIK it's not huge for macOS at the moment.

**Comment by CodingMate** [comment]
My main concern isn't the size, it's the fact that Apple will not allow the app to run unless it's signed.
Currently, Apple only allows apps to be executed as long when they are downloaded from AppStore or they are identified/signed by identified developers, Otherwise, the app will be blocked from running.
The workaround is to unblock the app from the settings, however, if you don't use the self-contained app, you will have to unblock each dll/so file you have in your app.

**Comment by agocke** [comment]
I'm no Mac notarization expert, but I thought that was what https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_disable-library-validation was for?

**Comment by CodingMate** [comment]
That link will disable the validation for the libraries and plug-ins only, it will not disable the app validation.
That entitlement will be really helpful after signing the app to disable validating the dlls/so files if we’re not publishing a single-file.

**Comment by Meowzz95** [comment]
Is there a workaround in .NET 5? I'm publishing single file using .NET 5 RC but still get the same error when signing the binary.

**Comment by richardlalancette** [comment]
Guys. This is a make or break for my team. If this cannot be smoothed out, I might have to go for another ecosystem, and I would hate to have to give away .net core.

I'm going to be using .net core to build command line art processing tools, for our designer workflow and this really needs to get addressed as I will be distributing my tools to mac, windows and maybe linux users.

**Comment by agocke** [comment]
We do plan to fix this, and it's high priority for .NET 6, unfortunately we just haven't had time to fix it for 5.0. As @swaroop-sridhar mentioned, you can use a macOS app bundle to distribute a multi-file app that appears as a single file to consumers.

Unfortunately, command line apps are one of the tools that do not normally use macOS bundles. Until we can fix this authoring issue, it won't be possible to create single-file command line apps on macOS.

**Comment by Meowzz95** [comment]
@richardlalancette 
I know this is very important when deploying CLI to customers/users, and I tried countless method without any luck. Until one day I thought why not put the effort into providing an easier way for the user to bypass gatekeeper! You can write a "run script" to use `spctl` to whitelist the files/dylibs in single-file output and then execute the CLI tool. This way, at least, the user can run the tool easily, well, easier.

**Comment by richardlalancette** [comment]
Oh, this is a risky business @Meowzz95, but nonetheless a good lead. Thank you.

This can't be the final solution though.

**Comment by sttz** [comment]
I'm using [CoreRT](https://github.com/dotnet/corert) for command line tools on macOS. It AOT-compiles to a single binary that can be signed and notarized.

CoreRT is experimental and has a bumpy future, so using it comes with some risk and you might not be able to use it with the latest .Net releases and features. But right now it's the only single-file solution I can use on macOS.

**Comment by Meowzz95** [comment]
@richardlalancette Agreed that .net team should fix this. But using `spctl` to whitelist ONLY your application files, don't turn off the gatekeeper and it should be just fine, as least same as user manually trust the binary.

**Comment by richardlalancette** [comment]
OH I see! I didn't realise specific files could be changed. Well worth a quick research spike :) thank you. @Meowzz95

**Comment by Meowzz95** [comment]
@richardlalancette HAHA, yes we can label specific files and whitelist them in gatekeeper, I'll show you my solution here.

```
echo "adding x permission to the program, you may need to enter sudo password or use biometric authentication"
sudo chmod +x YOUR_BINARY_EXEC
echo "adding exceptions to gatekeeper...you may need to enter sudo password or use biometric authentication"
ls | xargs -I {} echo $(pwd)/{}|  xargs -I {} spctl --add --label "SOME_LABEL" {}
spctl --enable --label "SOME_LABEL"
echo "launching YOUR APP..."
$(pwd)/YOUR_BINARY_EXEC
```

This is a very straight forward solution without much consideration on skipping the `chomd` and `spctl` on the second run, you can add logic to only run those in the first run so that user does not need to enter the password every time they use the app.

**Comment by richardlalancette** [comment]
@Meowzz95 Very generous of you to share your results. Thank you!

**Comment by vitek-karas** [comment]
Related work in mono: https://github.com/mono/mono/issues/17881

**Comment by VSadov** [comment]
Fixed in https://github.com/dotnet/runtime/pull/46558

**Comment by Jmales** [comment]
Hey guys. Not sure how to solve my problem, even after @VSadov fix.

I have a singleFile bundle inside another App and when I sign the whole contents of the App it always fails in the singleFile bundle binary with the mentioned error. I'm failing to understand how to fix the issue.

Can someone help?

**Comment by vitek-karas** [comment]
As mentioned above by @VSadov this has been fixed for .NET 6 here: https://github.com/dotnet/runtime/pull/46558. As far as I know we don't have plans to try to backport this to .NET 5.

**Comment by Jmales** [comment]
This single-file was my workaround for this: https://github.com/dotnet/runtime/issues/46990

Everything works awesome, except for the signing of the SingleFile binary/bundle. Now we have no idea on how to tackle the problem at hands.

**Comment by VSadov** [comment]
@Jmales - I am a bit confused. Did you have a problem signing a .NET 6.0 singlefile app? 

I wonder what was the SDK build number. The fix is fairly recent and it takes some time for a fix to make it from the development branch to the published installers. Maybe the fix was not there yet.

**Comment by richardlalancette** [comment]
Really looking for this. This was a pretty major roadblock for my team.

**Comment by VSadov** [comment]
Recent 6.0 and 6.0.preview1 SDK builds should contain the fix. 
@richardlalancette - could you check if the fix is sufficient for your scenario?

**Comment by richardlalancette** [comment]
I'm still on .net 5.0 here.

Is the 6.0 preview SDK good enough for production apps?
